### PR TITLE
Add support for Inventec d7032q28b.  Update Inventec d7032q28b.

### PR DESCRIPTION
**Pull request by ghjoec**
Add support for Inventec d7032q28b.

### CODE REVIEW

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/modules/builds/gpio-ich.c [hunk 1]
```diff
@@ -0,0 +1,548 @@
+/*
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/modules/builds/gpio-ich.c]: Please don't include copies of existing kernel modules in your platform. 
You can enable CONFIG_GPIO_ICH=m in the 3.16 kernel config if its not currently available.

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/auto/x86_64_inventec_d7032q28b.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+###############################################################################
+#
+# x86_64_inventec_d7032q28b Autogeneration Definitions.
+#
+###############################################################################
+
+cdefs: &cdefs
+- x86_64_inventec_d7032q28b_CONFIG_INCLUDE_LOGGING:
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/auto/x86_64_inventec_d7032q28b.yml]: All of these config defines are by convention uppercase. 
Please s/x86_64_inventec_d7032q28b_CONFIG/X86_64_INVENTEC_D7032Q28B_CONFIG/g and regenerate.

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/debug.c [hunk 3]
```diff
@@ -0,0 +1,45 @@
+#include "x86_64_inventec_d7032q28b_int.h"
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/debug.c]: Please remove this file.

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+static int
+_onlp_fani_info_get_fan(int local_id, onlp_fan_info_t* info)
+{
+    int   fd, len, nbytes = 10;
+    char  r_data[10]   = {0};
+    char  fullpath[65] = {0};
+
+    /* check if fan is present
+     */
+    sprintf(fullpath, "%s%s", PREFIX_PATH_ON_MAIN_BOARD, fan_path[local_id].present);
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c]: Please remote the OPEN_READ_FILE() macros and replace with native calls to the onlp_file_* APIs only.

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+/*
+ * This function sets the speed of the given fan in RPM.
+ *
+ * This function will only be called if the fan supprots the RPM_SET
+ * capability.
+ *
+ * It is optional if you have no fans at all with this feature.
+ */
+int
+onlp_fani_rpm_set(onlp_oid_t id, int rpm)
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c]: Remove function.

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+        case FAN_3_ON_MAIN_BOARD:
+        case FAN_4_ON_MAIN_BOARD:
+        case FAN_5_ON_MAIN_BOARD:
+        case FAN_6_ON_MAIN_BOARD:
+            sprintf(fullpath, "%s%s", PREFIX_PATH_ON_MAIN_BOARD, fan_path[local_id].ctrl_speed);
+            break;
+        default:
+            return ONLP_STATUS_E_INVALID;
+    }
+    sprintf(data, "%d", p);
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c]: These can all be replaced with a call onlp_file_write_int()

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+
+/*
+ * This function sets the fan speed of the given OID as per
+ * the predefined ONLP fan speed modes: off, slow, normal, fast, max.
+ *
+ * Interpretation of these modes is up to the platform.
+ *
+ */
+int
+onlp_fani_mode_set(onlp_oid_t id, onlp_fan_mode_t mode)
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c]: Remove

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+/*
+ * This function sets the fan direction of the given OID.
+ *
+ * This function is only relevant if the fan OID supports both direction
+ * capabilities.
+ *
+ * This function is optional unless the functionality is available.
+ */
+int
+onlp_fani_dir_set(onlp_oid_t id, onlp_fan_dir_t dir)
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c]: Remove

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+{
+    return ONLP_STATUS_E_UNSUPPORTED;
+}
+
+/*
+ * Generic fan ioctl. Optional.
+ */
+int
+onlp_fani_ioctl(onlp_oid_t id, va_list vargs)
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/fani.c]: Remove

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/ledi.c [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+    local_id = ONLP_OID_ID_GET(id);
+    		
+    /* get fullpath */
+    sprintf(fullpath, "%s%s/%s", prefix_path, last_path[local_id], filename);
+		
+	/* Set the onlp_oid_hdr_t and capabilities */
+    *info = linfo[ONLP_OID_ID_GET(id)];
+
+    /* Set LED mode */
+    if (deviceNodeReadString(fullpath, data, sizeof(data), 0) != 0) {
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/ledi.c]: onlp_file_read_str()

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/ledi.c [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+{
+    int  local_id;
+    char fullpath[50] = {0};		
+
+    VALIDATE(id);
+	
+    local_id = ONLP_OID_ID_GET(id);
+    sprintf(fullpath, "%s%s/%s", prefix_path, last_path[local_id], filename);	
+    
+    if (deviceNodeWriteInt(fullpath, onlp_to_driver_led_mode(local_id, mode), 0) != 0)
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/ledi.c]: onlp_file_write_int()

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/ledi.c [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+    }
+
+    return ONLP_STATUS_OK;
+}
+
+/*
+ * Generic LED ioctl interface.
+ */
+int
+onlp_ledi_ioctl(onlp_oid_t id, va_list vargs)
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/ledi.c]: Remove

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/platform_lib.c [hunk 13]
```diff
@@ -0,0 +1,200 @@
+/************************************************************
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/platform_lib.c]: I understand that this was copied from an existing platform implementation. 
Please do not propagate these custom interfaces but remove them and use the existing common onlp_file_* APIs.

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/psui.c [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+    *value = 0;
+
+    if (PSU1_ID == id) {
+        sprintf(node_path, "%s%s", PSU1_AC_HWMON_PREFIX, node);
+    }
+    else if (PSU2_ID == id) {
+        sprintf(node_path, "%s%s", PSU2_AC_HWMON_PREFIX, node);
+    }
+    
+    ret = deviceNodeReadString(node_path, buf, sizeof(buf), 0);
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/psui.c]: onlp_fiel_read_str()

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/sfpi.c [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+ 4,  5,  6,  7,  9,  8, 11, 10,
+ 0,  1,  2,  3, 12, 13, 14, 15,
+16, 17, 18, 19, 28, 29, 30, 31,
+20, 21, 22, 23, 24, 25, 26, 27
+};
+
+#define FRONT_PORT_TO_MUX_INDEX(port) (sfp_mux_index[port]+MUX_START_INDEX)
+
+static int
+as7512_32x_sfp_node_read_int(char *node_path, int *value, int data_len)
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/sfpi.c]: Please clean up your copy-paste from the AS7512.

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/modules/builds/gpio-ich.c (hunk 1 above)
Comment (ghjoec) [reply packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/modules/builds/gpio-ich.c]: Thanks for reviewing.  Got it.  I have a few questions.
Changing the kernel config will affect all platforms? Is this what we want?
If I want to modify kernel source, I will need to create a patch, right? Does that means my patch will be applied for all platforms too? Can we select the patches to apply for a particular platform?
If our platform kernel module under packages/platforms/<vendor>/x86-64/x86-64-<vendor>-<platform>/modules/builds/  consists of multiple .c files? How do I modify the Makefile? I don't see any example of this.  It seems to me each kernel module can only have one .c in ONL.  Can I put their .h files in the same directory as .c files?  Thanks in advance.

**Comment by ghjoec** [comment]
Just submitted another commit "Update Inventec d7032q28b.",  commit#97247e2 to fix the previous commit.  Unable to create another pull request, now it's under the same pull request#252. Please review.

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/platform_lib.c [hunk 16]
```diff
@@ -30,9 +30,13 @@
...
 #include <fcntl.h>
 #include <AIM/aim.h>
+#include <onlplib/file.h>
+#include <onlp/onlp.h>
 #include "platform_lib.h"
 
-int deviceNodeWrite(char *filename, char *buffer, int buf_size, int data_len)
+#define PSU_NODE_MAX_PATH_LEN 64
+
+int _onlp_file_write(char *filename, char *buffer, int buf_size, int data_len)
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/platform_lib.c]: You don't need all of the onlp_file_{write,read} implementations. Perhaps my previous comments were too brief. 

Please use the exisiting ones already available in the onlplib module:
https://github.com/opencomputeproject/OpenNetworkLinux/blob/master/packages/base/any/onlp/src/onlplib/module/inc/onlplib/file.h

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/psui.c [hunk 17]
```diff
@@ -58,7 +58,7 @@ psu_status_info_get(int id, char *node, int *value)
         sprintf(node_path, "%s%s", PSU2_AC_HWMON_PREFIX, node);
     }
     
-    ret = deviceNodeReadString(node_path, buf, sizeof(buf), 0);
+    ret = onlp_file_read_string(node_path, buf, sizeof(buf), 0);
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/psui.c]: Can be replaced directly with a call to onlp_file_read_int()

File: packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/psui.c [hunk 18]
```diff
@@ -83,7 +83,7 @@ psu_ym2651_pmbus_info_get(int id, char *node, int *value)
         sprintf(node_path, "%s%s", PSU2_AC_PMBUS_PREFIX, node);
     }
 
-    ret = deviceNodeReadString(node_path, buf, sizeof(buf), 0);
+    ret = onlp_file_read_string(node_path, buf, sizeof(buf), 0);
```
Comment (jnealtowns) [inline packages/platforms/inventec/x86-64/x86-64-inventec-d7032q28b/onlp/builds/src/module/src/psui.c]: onlp_file_read_int()

**Comment by jnealtowns** [comment]
We're close on this. Thanks for you submission. Just a few more minor changes required (mostly porting the the existing FILE apis)

**Comment by ghjoec** [comment]
Closing this request and issuing new one.

**Pull request closed without merging**